jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py

```diff
@@ -1,128 +1,271 @@
+    def _init_empty_index(self):
+        self._index.init_index(
+            max_elements=self.max_elements,
+            ef_construction=self.ef_construction,
+            M=self.max_connection,
+        )
+        self._ids_to_inds = bidict()
 
     @requests(on='/search')
-    def search(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+    def search(
+        self, docs: Optional[DocumentArray] = None, parameters: Dict = {}, **kwargs
+    ):
+        """Attach matches to the Documents in `docs`, each match containing only the
+        `id` of the matched document and the `score`.
+
+        :param docs: An array of `Documents` that should have the `embedding` property
+            of the same dimension as vectors in the index
+        :param parameters: Dictionary with optional parameters that can be used to
+            override the parameters set at initialization. Supported keys are
+            `traversal_paths`, `top_k` and `ef_query`.
         """
+        if docs is None:
+            return
 
-        :param docs: `Document` with `.embedding` the same shape as the `Documents`
-            it has stored.
-        :param parameters: dictionary to define the `traversal_paths` and the `top_k`.
+        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
+        docs_search = docs.traverse_flat(traversal_paths)
+        if len(docs_search) == 0:
+            return
+
+        ef_query = parameters.get('ef_query', self.ef_query)
+        top_k = parameters.get('top_k', self.top_k)
+
+        self._index.set_ef(ef_query)
+
+        if top_k > len(self._ids_to_inds):
+            self.logger.warning(
+                f'The `top_k` parameter is set to a value ({top_k}) that is higher than'
+                f' the number of documents in the index ({len(self._ids_to_inds)})'
+            )
+            top_k = len(self._ids_to_inds)
+
+        embeddings_search = docs_search.embeddings
+        if embeddings_search.shape[1] != self.dim:
+            raise ValueError(
+                'Query documents have embeddings with dimension'
+                f' {embeddings_search.shape[1]}, which does not match the dimension of'
+                f' the index ({self.dim})'
+            )
 
-        :return: Attaches matches to the Documents sent as inputs, with the id of the
-            match, and its embedding.
+        indices, dists = self._index.knn_query(docs_search.embeddings, k=top_k)
+
+        for i, (indices_i, dists_i) in enumerate(zip(indices, dists)):
+            for idx, dist in zip(indices_i, dists_i):
+                match = Document(id=self._ids_to_inds.inverse[idx])
+                match.scores[self.metric] = dist
+
+                docs_search[i].matches.append(match)
+
+    @requests(on='/index')
+    def index(
+        self, docs: Optional[DocumentArray] = None, parameters: Dict = {}, **kwargs
+    ):
+        """Index the Documents' embeddings. The documents should not be already
```

Comment:
What happens if I send a doc with the id of a previously indexed doc? This could happen quite easily and should be covered. Ignoring it, raising a warning etc.

Reply:
You are not missing anything - I have conciously decided against it.

Filling the document with attributes is the responsibility of the storage indexer. The search indexer is not even guaranteed to have the embedding (e.g. PQ Faiss), and may not be optimized for its retrieval even if it has it.

Reply:
The issue I see here is that this involves checking that id is not already in the index for each item that we want to index. Even though this is executed in constant time (relative to size of index), if we index say 1 million documents, it could still add up (I calculated that it would add 10 seconds). If that's a delay we are willing tolerate, a check can be added

Reply:
I was wrong, it would be only 0.1 seconds. Will change the logic so that items with existing id will get updated (easiest to implement and one that makes most sesnse imo)

Reply:
> items with existing id will get updated 

But a raise warning at the end with the list of ids that were updated.

Reply:
Currently it is already logging each such case (as info, not warning)

https://github.com/jina-ai/executors/blob/26cb8de69e514b6c83d7b5993063b125a588fa11/jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py#L172-L176

Do you think it needs to be a warning? I put it as info because I think it is most likely what the user actually intended (because overlapping ids have to be created explicitly)